### PR TITLE
Blocks: Add quote blocks style controls

### PR DESCRIPTION
**Pull request by youknowriad**
#311 

In this PR I'm adding the quote styles controls.

The question here is how we provide these styles to the theme developers, should they be automatically included, or do we expect theme authors to implement these two styles on their own? This question could probably be answered separately.

### CODE REVIEW

File: blocks/library/quote/index.js [hunk 1]
```diff
@@ -20,11 +20,21 @@ registerBlock( 'core/quote', {
 		citation: html( 'footer' )
 	},
 
+	controls: [ '1', '2' ].map( ( level ) => ( {
```
Comment (aduth) [inline blocks/library/quote/index.js]: Not sure `level` applies as well here as it does for a heading. `variant`, `variation`, or `style` might be more accurate.

File: blocks/library/quote/index.js [hunk 2]
```diff
@@ -20,11 +20,21 @@ registerBlock( 'core/quote', {
...
+		icon: 'format-quote',
+		title: wp.i18n.sprintf( wp.i18n.__( 'Quote %s' ), level ),
```
Comment (aduth) [inline blocks/library/quote/index.js]: We should probably use `%d` placeholder, see also https://github.com/WordPress/gutenberg/pull/423#discussion_r112088447

File: blocks/library/quote/index.js (hunk 1 above)
Comment (aduth) [inline blocks/library/quote/index.js]: Minor: Do you think it's any different or better to treat `'1'` as a string vs. `1` as a number?

**Comment by aduth** [comment]
>The question here is how we provide these styles to the theme developers, should they be automatically included, or do we expect theme authors to implement these two styles on their own? 

Having a hard time tracking down the specific discussion, but I know this had been raised at some point.

Maybe a theme can define compatibility via [`add_theme_support`](https://developer.wordpress.org/reference/functions/add_theme_support/)? In which case we'd need a way to vary client-side block controls by server-side theme supports. Might tie in desire outlined in #399 to have server-side block filtering.

File: blocks/library/quote/index.js (hunk 1 above)
Comment (youknowriad) [inline blocks/library/quote/index.js]: I think a string is better to avoid any confusion while parsing/serializing the attribute back.

Comment (aduth) [inline blocks/library/quote/index.js]: Oh, that's a good point. Speaking of... how are we parsing the attribute?

File: blocks/library/quote/index.js [hunk 3]
```diff
@@ -20,11 +20,21 @@ registerBlock( 'core/quote', {
...
 	},
 
+	controls: [ '1', '2' ].map( ( variation ) => ( {
+		icon: 'format-quote',
+		title: wp.i18n.sprintf( wp.i18n.__( 'Quote %d' ), variation ),
+		isActive: ( { style = '1' } ) => style === style,
+		onClick( attributes, setAttributes ) {
+			setAttributes( { style: variation } );
+		},
+		variation
```
Comment (aduth) [inline blocks/library/quote/index.js]: Looks like we've highlighted a problem that our `<Toolbar />` is tightly coupled to the idea of a heading level:

https://github.com/WordPress/gutenberg/blob/d068400/editor/components/toolbar/index.js#L24

Comment (youknowriad) [inline blocks/library/quote/index.js]: Do you think we should rename it `variation` or `suffix` or something like that?

Comment (aduth) [inline blocks/library/quote/index.js]: A shame that our [blueprinting](https://wpcoredesign.mystagingwebsite.com/gutenberg/#blueprints) wordsmith @jasmussen is absent today 😄

File: blocks/library/quote/index.js (hunk 2 above)
Comment (youknowriad) [inline blocks/library/quote/index.js]: revert this change because I'm using strings

File: blocks/library/quote/index.js (hunk 3 above)
Comment (youknowriad) [inline blocks/library/quote/index.js]: I like subscript

**Comment by nylen** [comment]
The quote style is a number; let's treat it like a number.

> I think a string is better to avoid any confusion while parsing and serializing the attribute back.

To me a better way to address this is to cast the value where it may be necessary to do so.  Done in 968ec56.

I also changed `Quote %d` to `Quote style %d` because the latter makes more sense without context.

**Review by nylen — APPROVED** [review]

**Review by aduth — DISMISSED** [review]
Still looks like we're not ever parsing the style attribute.

**Comment by youknowriad** [comment]
Thanks @nylen for these changes, I do prefer the simpler string parsing, the 1,2 here are just identifiers for me and not really numbers but It's really minor. Let's ship it.

File: blocks/library/quote/index.js [hunk 4]
```diff
@@ -17,14 +17,30 @@ registerBlock( 'core/quote', {
 
 	attributes: {
 		value: query( 'blockquote > p', html() ),
-		citation: html( 'footer' )
+		citation: html( 'footer' ),
+		style: node => {
+			const value = attr( 'blockquote', 'class' )( node );
+			const match = value.match( /\bblocks-quote-style-(\d+)\b/ );
```
Comment (aduth) [inline blocks/library/quote/index.js]: What happens if the `class` is removed from the `blockquote` ?

**Comment by aduth** [comment]
Makes me wonder at what point we start to defer to comment-serialized attributes as hints for what also exists in the markup, just to avoid complexity of matching class patterns etc.

Which leads into another question... if we _did_ serialize the quote style in the comment, it'd have to be parsed as a string, yeah? Unless we did adopt JSON or another pattern to distinguish between strings and other types:

`style` as string:

```html
<!-- wp:quote style:2 -->
```

`style` as number:

```html
<!-- wp:quote foo:"definitely-a-string" style:2 -->
```
```html
<!-- wp:quote {"foo":"definitely-a-string","style":2} -->
```

**Comment by youknowriad** [comment]
Good question @aduth I guess like HTML attributes comment attributes are strings by default. I guess everything could be represented as a string but JSON attributes could be a good alternative especially for complex attributes (object attributes).

Note that we already store the "align" attribute for the text block in the comment

**Comment by aduth** [comment]
We could adopt strict JSON behavior for the value, always wrapping string in `"`. Also resolves the need for values with spaces.

http://www.json.org/value.gif

**Comment by nylen** [comment]
Following up on a couple of things here...

> I do prefer the simpler string parsing, the 1,2 here are just identifiers for me and not really numbers

In this case I think the right thing to do would be to make them identifiers (string names) rather than numbers.  I don't have a strong preference, I just think we should decide one way or the other and then stick to that decision fully.

> Makes me wonder at what point we start to defer to comment-serialized attributes as hints for what also exists in the markup, just to avoid complexity of matching class patterns etc.

I think we should try to avoid these attributes as much as possible, again in the interest of avoiding duplicated information.

> Which leads into another question... if we did serialize the quote style in the comment, it'd have to be parsed as a string, yeah? Unless we did adopt JSON or another pattern to distinguish between strings and other types:

Agree that our current method of storing attributes isn't very robust.  I'd like to see either HTML-style attributes `name="value"` or - if we know that we really need to store typed data - this option:

```html
<!-- wp:quote {"foo":"definitely-a-string","style":2} -->
```

**Comment by jasmussen** [comment]
> In this case I think the right thing to do would be to make them identifiers (string names) rather than numbers. I don't have a strong preference, I just think we should decide one way or the other and then stick to that decision fully.

I will defer to you all on what exactly to do here, as I'm sure there are intricate reasons one way or the other. To add some context, we are likely to see many other blocks as plugins, including special quote blocks perhaps with extra parameters. So when is it a new block vs. when is it a _style_ for an existing block? I think it should be a style when it's the exact same markup and exact same surfaced parameters, and it would be lovely if _someday maybe_ plugins could also add extra styles to existing blocks. That might be more scalable if they were numbers.

**Comment by nylen** [comment]
> [plugins adding extra styles] might be more scalable if they were numbers.

I think the opposite is true.  How do we guarantee that the numbers wouldn't conflict or change over time?  It would be pretty bad to have a quote's style change depending on which plugins happened to be active at the time, for example.

Here are the pros and cons I can think of at the moment.

### Numbers

**Pros**: Simpler mapping between data / names / toolbar icons; we don't have to try to name these things.

**Cons**: More difficult for plugin authors.  As above, how do we guarantee that numbers don't conflict or change over time?

### Names

**Pros**: More explicit, more opportunity for plugins to create their own styles (they should probably be namespaced by plugin name, with a standard way to add styles across blocks).

**Cons**: We have to name them; we have to surface the names in a user-friendly and accessible way.

Overall I think names is probably a better way to go, but this can wait a while and we need to think through what we want this functionality to be able to do.  For example, anything we allow plugins to do [should be namespaced](https://nacin.com/2010/05/11/in-wordpress-prefix-everything/).
